vega/utils.py

```diff
@@ -30,22 +36,61 @@ def get_content(path):
         return f.read()
 
 
-def sanitize(value):
-    """Make sure that the data is compatible with JSON."""
-    if pd.isnull(value):
-        # json doesn't support nan
-        return None
-    return value
+def escape(string):
+    """Escape the string."""
+    return cgi.escape(string, quote=True)
 
 
-def data(data, columns):
-    """Create a dictionary from a pandas data frame."""
-    res = []
-    for row in data.tolist():
-        res.append({k: sanitize(v) for k, v in zip(columns, row)})
-    return res
+def sanitize_dataframe(df):
+    """Sanitize a DataFrame to prepare it for serialization.
+    
+    * Make a copy.
+    * Raise ValueError if it has a hierarchical index.
+    * Convert categoricals to strings.
+    * Convert datetimes to strings.
+    * Convert floats to objects and replace NaNs by None.
+    """
+    import pandas as pd
+    df = df.copy()
+    
+    if isinstance(df.index, pd.core.index.MultiIndex):
+        raise ValueError('Hierarchical indices not supported')
+    if isinstance(df.columns, pd.core.index.MultiIndex):
+        raise ValueError('Hierarchical indices not supported')
 
+    for col_name, dtype in df.dtypes.iteritems():
+        if str(dtype) == 'category':
+            # XXXX: work around bug in to_json for categorical types
+            # https://github.com/pydata/pandas/issues/10778
+            df[col_name] = df[col_name].astype(str)
+        if str(dtype).startswith('datetime'):
+            # Convert datetimes to strings
+            df[col_name] = df[col_name].astype(str).replace('NaT','')
+        if str(dtype).startswith('float'):
+            # For floats, convert nan->None
+            s = df[col_name].astype(object)
+            df[col_name] = s.where(s.notnull(), None)
+    return df
 
-def escape(string):
-    """Escape the string."""
-    return cgi.escape(string, quote=True)
+
+def prepare_spec(spec, data=''):
+    """Prepare a Vega-Lite spec for sending to the frontend.
+    
+    This allows data to be passed in either as part of the spec
+    or separately. If separately, the data is assumed to be a
+    pandas DataFrame or object that can be converted to to a DataFrame.
+    """
+    import pandas as pd
+
+    # No data provided
+    if data is '' and 'data' not in spec:
+        raise ValueError('No data provided')
+    # Data provided, prepared it and put it into the spec
+    if not data is '':
```

Comment:
probably ``==`` rather than ``is`` here. I'm not sure if the empty string is a singleton.